functorch/_src/aot_autograd.py

```diff
@@ -1109,6 +1112,45 @@ def remove_dupe_args(args):
     def add_dupe_args(args):
         return [args[add_dupe_map[i]] for i in range(duped_arg_len)]
 
+    def maybe_wrap_debug(f):
+        if not config.debug_assert:
+            return f
+
+        @wraps(f)
+        def debug_wrapper(*args):
+            # Test that the computed remove/add arg functions are an inverse
+            new_args = add_dupe_args(remove_dupe_args(args))
+            seen = {}
+            for i, (x, y) in enumerate(zip(new_args, args)):
+                seen[y] = None
+                assert x is y, (
+                    "At compilation time, this graph was compiled under the "
+                    "assumption that some inputs were duplicate, but at runtime "
+                    f"input {i} was not a duplicate of {add_dupe_map[i]}.  " +
+                    GUARD_BUG_BOILERPLATE
+                )
+            # This is only an error if there is metadata mutation on both of
+            # the duped arguments; in this case, we need to know what order
+            # the metadata mutation applies in.  You'll get the correct result
+            # otherwise, because a graph that assumes distinct inputs works if
+            # you dupe the inputs (the gradient contributions from each input
+            # will get summed up appropriately.)
+            """
+            assert len(seen) == unique_args, (
+                "At compilation time, this graph was compiled under the assumption "
+                f"that there would be {unique_args} distinct arguments, but at "
+                f"runtime there were only {len(seen)} distinct arguments.  " +
+                GUARD_BUG_BOILERPLATE
+            )
+            """
+            return f(*args)
+
+        return debug_wrapper
+
+    # Fastpath
+    if not dropped_args:
```

Comment:
Doesn't that mean that when we have no dupe args, we'll always skip our debug checks about requires-gradness (and eventually input alias info?)